internal/collector/processor/nodebatcher/node_batcher.go

```diff
@@ -135,154 +130,83 @@ func (b *batcher) genBucketID(node *commonpb.Node, resource *resourcepb.Resource
-
-		dead:            uint32(0),
-		cyclesUntouched: uint32(0),
-		parent:          parent,
+) *nodeBatch {
+	return &nodeBatch{
+		parent:   parent,
+		format:   format,
+		node:     node,
+		resource: resource,
+		items:    make([]*tracepb.Span, 0, initialBatchCapacity),
 	}
-
-	return nb
 }
 
-func (nb *nodeBatcher) add(spans []*tracepb.Span) {
-	// Reset cycles untouched to 0
-	atomic.StoreUint32(&nb.cyclesUntouched, 0)
-	// Add will always add these spans to the current batch, if ok is false it means that
-	// we need to cut
-	var b *batch
-	closed := true
-	cutBatch := false
-	for closed {
-		// atomic.LoadPointer only takes unsafe.Pointer interfaces. We do not use unsafe
-		// to skirt around the golang type system.
-		b = (*batch)(atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&nb.currBatch))))
-		cutBatch, closed = b.add(spans)
-	}
-	if atomic.LoadUint32(&nb.dead) == nodeStatusDead || cutBatch {
-		statsTags := processor.StatsTagsForBatch(
-			nb.parent.name, processor.ServiceNameForNode(nb.node), nb.spanFormat,
-		)
-		if cutBatch {
-			stats.RecordWithTags(context.Background(), statsTags, statBatchSizeTriggerSend.M(1))
-		} else {
-			stats.RecordWithTags(context.Background(), statsTags, statBatchOnDeadNode.M(1))
-		}
-		// Shoot off event so that we don't block this add on other adds
-		go nb.cutBatch(b)
-	}
-}
+func (nb *nodeBatch) add(spans []*tracepb.Span) {
+	nb.mu.Lock()
+	nb.items = append(nb.items, spans...)
+	nb.cyclesUntouched = 0
 
-// cutBatch cuts the provided batch, and sets a new batch on this nodeBatcher
-func (nb *nodeBatcher) cutBatch(b *batch) {
-	initialCap := b.getCurrentCap()
-	currSize := b.getCurrentItemCount()
-	reducedCap := initialCap >> 1
-	for currSize < reducedCap && reducedCap > initialBatchCapacity {
-		initialCap = reducedCap
-		reducedCap = reducedCap >> 1
-	}
-	// atomic.CompareAndSwapPointer only takes unsafe.Pointer interfaces. We do not use unsafe
-	// to skirt around the golang type system.
-	currBatchPtr := (*unsafe.Pointer)(unsafe.Pointer(&nb.currBatch))
-	swapped := atomic.CompareAndSwapPointer(
-		currBatchPtr,
-		unsafe.Pointer(b),
-		unsafe.Pointer(newBatch(initialCap, nb.sendBatchSize)),
-	)
-	// Since we are doing an atomic compare and swap, this batch will only be sent once.
-	if swapped {
-		b.closeBatch()
-		atomic.StoreInt64(&nb.lastSent, time.Now().UnixNano())
-		nb.sendBatch(b)
+	var itemsToProcess []*tracepb.Span
+	if uint32(len(nb.items)) > nb.parent.sendBatchSize || nb.dead == nodeStatusDead {
+		itemsToProcess = nb.items
+		nb.items = make([]*tracepb.Span, 0, len(nb.items))
+		nb.lastSent = time.Now().UnixNano()
 	}
-}
+	nb.mu.Unlock()
 
-func (nb *nodeBatcher) sendBatch(batch *batch) {
-	spans := batch.getSpans()
-	statsTags := processor.StatsTagsForBatch(
-		nb.parent.name, processor.ServiceNameForNode(nb.node), nb.spanFormat,
-	)
-	stats.RecordWithTags(context.Background(), statsTags, statBatchSize.M(int64(len(spans))))
+	if len(itemsToProcess) > 0 {
```

Comment:
In order to not penalize on goroutine for this probably we should put this into a channel and another goroutine should pick this up. Maybe for the moment a TODO and a followup PR is fine.

Reply:
added a todo.

Reply:
I'm not so sure: perhaps we delegate that to the next element in the pipeline: likely a queue that is already putting stuff via channel.